# `RefCell<T>`与内部可变性模式

*内部可变性*interior mutability 是 Rust 的一种设计模式，它让你修改数据即使数据有一个不可变引用；通常来说这是不被借用规则所允许的。为了修改数据，这个模式在数据结构中使用了`unsafe`代码用于绕过 Rust 的通常的管理可变性和借用规则。第十九章将会将到不安全代码。当我们能确保借用规则能在运行被遵循时，我们可以使用用到了内部可变性模式的类型，即便编译器不能保证它。涉及到的`unsafe`代码被封装在安全的 API 中，我们外部的类型依旧是不可变的。

## 通过`RefCell<T>`在运行时遵循借用规则

不同于`Rc<T>`，`RefCell<T>`类型代表一个数据的唯一所有权。是什么让`RefCell<T>`与`Box<T>`不同？我们回想一下第四章的借用规则：

- 任何时候，你可以有一个借用引用或者若干不可变引用，且两者不能同时存在。
- 引用必须总是有效的。

对于引用和`Box<T>`，引用规则不可变性作用于编译时期。对于`RefCell<T>`，不可变性作用于运行时期。对于引用，如果你打破规则，则会得到编译错误。对于`RefCell<T>`，如果你打破规则，则会得到 panic 并强制退出。

在编译时检查借用规则的好处是错误能在开发时就被捕获，并且对运行性能没有影响因为所有的分析都被完成了。由于这些原因，对于多数情况而言，编译期的借用规则检查是最好的选择，这也是 Rust 的默认选择。

在运行时检查借用规则的好处是允许特定的内存安全场景，即便它们是不被编译时期所允许的。静态分析，如 Rust 编译器，是天生保守的。有些代码的属性很难通过分析发现：最出名的例子是停机问题。

正因为游戏分析是不可能的，如果 Rust 编译器不能确保借用规则，它可能拒绝一个正常的程序；这种情况它是保守的。如果 Rust 接受一个不正常的程序，用户便不在信任 Rust 的决策。因此即使在编译器不能确保的情况下，但是你能确保你的代码能在运行时遵循借用规则时，`RefCell<T>`类型非常有用，

与`Rc<T>`类似，`RefCell<T>`只被允许与单线程场景，如果在多线程的上下文中出现则会出现编译错误。第十六章将会讲在多线程情况下怎么使用`RefCell<T>`的功能。

以下是选择`Box<T>`，`Rc<T>`或`RefCell<T>`的理由：

- `Rc<T>`允许一个数据拥有多个所有者；`Box<T>`和`RefCell<T>`只有唯一所有者。
- `Box<T>`在编译时允许不可变或可变借用；`Rc<T>`在编译时只允许不可变借用；`RefCell<T>`在运行时允许不可变或可变借用。
- `RefCell<T>`在运行时允许可变借用，我们可以修改`RefCell<T>`的内部值即使`RefCell<T>`是不可变的。

## 内部可变性：不可变值的可变借用

一个借用规则的推论是：当有一个不可变值，你不可以可变的借用它。例如以下代码就不可被编译：

```rust
fn main() {
    let x = 5;
    let y = &mut x;
}
```

报错如下：

```null
$ cargo run
   Compiling borrowing v0.1.0 (file:///projects/borrowing)
error[E0596]: cannot borrow `x` as mutable, as it is not declared as mutable
 --> src/main.rs:3:13
  |
2 |     let x = 5;
  |         - help: consider changing this to be mutable: `mut x`
3 |     let y = &mut x;
  |             ^^^^^^ cannot borrow as mutable

error: aborting due to previous error

For more information about this error, try `rustc --explain E0596`.
error: could not compile `borrowing`

To learn more, run the command again with --verbose.
```

然而，有一些情况，一个值能使用方法修改其自身同时对外保持不变性，会很有用。值方法外部的代码不能修改值。使用`RefCell<T>`就是获得内部可变性的一种方法。但是`RefCell<T>`并没有完全避开借用原则：编译器的借用检查器允许这样的内部可变性，但是会在运行时检查借用规则。如果违反了这些规则，会出现 panic 而不是编译错误。

### 案例：mock 对象

*测试替身*test double 是一个通用编程概念，即测试中一个类型替换另一个类型。*mock 对象*是特定类型的测试替身，其记录测试中发生的事以便断言正确的操作。

Rust 中的对象与其它语言的对象不一样，也没有在标准库中拥有 mock 对象的功能。然而，你可以创建一个结构体用作与 mock 对象一样的功能。

我们将要测试这样的场景：创建一个库用于追踪一个值与最大值的差别，并发送两者间差距的消息。例如，该库可以用于追踪用户调用了多少次 API 的数量限额。

该库只提供追踪与最大值的差值，以及何时发何种消息的功能。使用我们库的 app 需要提供发送消息机制：邮件、短信等。库本身不需要这些细节，仅需要实现`Message`特性：

```rust
pub trait Messenger {
    fn send(&self, msg: &str);
}

pub struct LimitTracker<'a, T: Messenger> {
    messenger: &'a T,
    value: usize,
    max: usize,
}

impl<'a, T> LimitTracker<'a, T>
where
    T: Messenger,
{
    pub fn new(messenger: &T, max: usize) -> LimitTracker<T> {
        LimitTracker {
            messenger,
            value: 0,
            max,
        }
    }

    pub fn set_value(&mut self, value: usize) {
        self.value = value;

        let percentage_of_max = self.value as f64 / self.max as f64;

        if percentage_of_max >= 1.0 {
            self.messenger.send("Error: You are over your quota!");
        } else if percentage_of_max >= 0.9 {
            self.messenger
                .send("Urgent warning: You've used up over 90% of your quota!");
        } else if percentage_of_max >= 0.75 {
            self.messenger
                .send("Warning: You've used up over 75% of your quota!");
        }
    }
}
```

上述代码中一个重要的部分是`Message`特性的`send`方法的入参是不可变引用的`self`和文字信息。这是我们 mock 对象所需要的接口。另一个重要的部分是我们希望在`LimitTracker`上测试`set_value`方法。我们可以改变`value`参数，但是`set_value`不会返回任何值用于断言。所以我们需要知道，使用实现了`Messenger`特性的值，以及一个`max`值创建了一个`LimitTracker`后，当我们传递不同`value`值时，消息发送者需要发送合适的消息。

我们需要一个 mock 对象，仅仅用于追踪信息，而不是在调用`call`后发邮件或短信。我们可以创建一个 mock 对象的实例，创建一个`LimitTracker`使用这个 mock 对象，调用`set_value`方法，检查 mock 对象的消息是否与预期一致。以下代码尝试实现一个 mock 对象，但是借用检查器通不过：

```rust
#[cfg(test)]
mod tests {
    use super::*;

    struct MockMessenger {
        sent_messages: Vec<String>,
    }

    impl MockMessenger {
        fn new() -> MockMessenger {
            MockMessenger {
                sent_messages: vec![],
            }
        }
    }

    impl Messenger for MockMessenger {
        fn send(&self, message: &str) {
            self.sent_messages.push(String::from(message));
        }
    }

    #[test]
    fn it_sends_an_over_75_percent_warning_message() {
        let mock_messenger = MockMessenger::new();
        let mut limit_tracker = LimitTracker::new(&mock_messenger, 100);

        limit_tracker.set_value(80);

        assert_eq!(mock_messenger.sent_messages.len(), 1);
    }
}
```

上述代码定义了一个拥有类型是`Vec<String>`的`sent_messages`字段的`MockMessenger`结构体。我们同样为其定义了一个关联函数`new`。接着为`MockMessenger`实现了`Messenger`特性，这样可以将其交给`LimitTracker`。在`send`方法定义中，我们把入参存放于`MockMessenger`中的`sent_messages`字段。

测试中，当`LimitTracker`被告知设置的`value`大于`max`值的百分之 75 时，我们需要知道什么会发生。首先我们创建一个拥有空信息的`MockMessenger`实例。接着创建一个拥有`mock_messenger`和`max`的`LimitTracker`实例。我们调用入参为 80 的`set_value`，这超过了最大值 100 的 75%。接着我们断言`mock_messenger`里的`sent_messages`字段有一条信息。

编译报错：

```null
$ cargo test
   Compiling limit-tracker v0.1.0 (file:///projects/limit-tracker)
error[E0596]: cannot borrow `self.sent_messages` as mutable, as it is behind a `&` reference
  --> src/lib.rs:58:13
   |
57 |         fn send(&self, message: &str) {
   |                 ----- help: consider changing this to be a mutable reference: `&mut self`
58 |             self.sent_messages.push(String::from(message));
   |             ^^^^^^^^^^^^^^^^^^ `self` is a `&` reference, so the data it refers to cannot be borrowed as mutable

error: aborting due to previous error

For more information about this error, try `rustc --explain E0596`.
error: could not compile `limit-tracker`

To learn more, run the command again with --verbose.
warning: build failed, waiting for other jobs to finish...
error: build failed
```

我们不可以改变`MockMessenger`用于追踪信息，因为`send`方法的入参是`self`的不可变引用。我们也不可以像错误信息里建议的那样使用`&mut self`，因为`send`的签名将不匹配`Messenger`特性的定义。

这种情况下内部可变性可以来帮忙！我们存储`sent_messages`于一个`RefCell<T>`，这样再调用`send`就可以改变`sent_messages`了：

```rust
#[cfg(test)]
mod tests {
    use super::*;
    use std::cell::RefCell;

    struct MockMessenger {
        sent_messages: RefCell<Vec<String>>,
    }

    impl MockMessenger {
        fn new() -> MockMessenger {
            MockMessenger {
                sent_messages: RefCell::new(vec![]),
            }
        }
    }

    impl Messenger for MockMessenger {
        fn send(&self, message: &str) {
            self.sent_messages.borrow_mut().push(String::from(message));
        }
    }

    #[test]
    fn it_sends_an_over_75_percent_warning_message() {
        // --snip--
        let mock_messenger = MockMessenger::new();
        let mut limit_tracker = LimitTracker::new(&mock_messenger, 100);

        limit_tracker.set_value(80);

        assert_eq!(mock_messenger.sent_messages.borrow().len(), 1);
    }
}
```

`sent_messages`字段现在的类型是`RefCell<Vec<String>>`而不是`Vec<String>`。在`new`函数中，我们现在创建了一个`RefCell<Vec<String>>`的空向量实例。

`send`方法的实现，第一个参数还是`self`的不可变借用，这样匹配特性的定义。我们调用`RefCell<Vec<String>>`的方法`borrow_mut`来获取一个其值可变引用，即字符串向量。接着在可变引用上调用`push`用于追踪测试里的信息。

最后一个改动是断言：为了看到内部向量中有多少项，我们调用`RefCell<Vec<String>>`的`borrow`方法来获取向量的不可变引用。

### 通过`RefCell<T>`在运行时追踪借用

我们分别用`&`和`&mut`语法来创建不可变和可变引用。使用`RefCell<T>`时，我们使用`borrow`和`borrow_mut`方法，它们皆是属于`RefCell<T>`的安全 API。`borrow`方法返回的是`Ref<T>`智能指针，而`borrow_mut`返回的是`RefMut<T>`智能指针。这两个类型都实现了`Deref`，因此我们可以像普通指针那样使用它们。

`RefCell<T>`持续追踪`Ref<T>`与`RefMut<T>`智能指针的数量。每次我们调用`borrow`，`RefCell<T>`增加不可变借用的计数。当一个`Ref<T>`离开作用域时，计数减一。与编译期的借用规则一样，`RefCell<T>`让我们在任何时刻都可以拥有多个不可变借用或者一个可变借用。

如果我们没有遵守这些规则，我们将得到的不是编译错误而是运行是的 panic。如下如果改动一下代码，尝试创建两个可变借用并在同一作用域中激活，`RefCell<T>`会在运行时期阻止我们：

```rust
    impl Messenger for MockMessenger {
        fn send(&self, message: &str) {
            let mut one_borrow = self.sent_messages.borrow_mut();
            let mut two_borrow = self.sent_messages.borrow_mut();

            one_borrow.push(String::from(message));
            two_borrow.push(String::from(message));
        }
    }
```

运行时报错：

```null
$ cargo test
   Compiling limit-tracker v0.1.0 (file:///projects/limit-tracker)
    Finished test [unoptimized + debuginfo] target(s) in 0.91s
     Running target/debug/deps/limit_tracker-d1b2637139dca6ca

running 1 test
test tests::it_sends_an_over_75_percent_warning_message ... FAILED

failures:

---- tests::it_sends_an_over_75_percent_warning_message stdout ----
thread 'main' panicked at 'already borrowed: BorrowMutError', src/lib.rs:60:53
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace


failures:
    tests::it_sends_an_over_75_percent_warning_message

test result: FAILED. 0 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out

error: test failed, to rerun pass '--lib'
```

...

## 结合`Rc<T>`和`RefCell<T>`的多所有者的可变数据

`RefCell<T>`的一种常见的用法是与`Rc<T>`结合。`Rc<T>`可以使一个数据拥有多个拥有者，但是仅允许不可变的访问。如果你有一个存储了`RefCell<T>`的`Rc<T>`，你可以获得一个多所有者并可变的值。

以 cons list 为例：

```rust
#[derive(Debug)]
enum List {
    Cons(Rc<RefCell<i32>>, Rc<List>),
    Nil,
}

use crate::List::{Cons, Nil};
use std::cell::RefCell;
use std::rc::Rc;

fn main() {
    let value = Rc::new(RefCell::new(5));

    let a = Rc::new(Cons(Rc::clone(&value), Rc::new(Nil)));

    let b = Cons(Rc::new(RefCell::new(3)), Rc::clone(&a));
    let c = Cons(Rc::new(RefCell::new(4)), Rc::clone(&a));

    *value.borrow_mut() += 10;

    println!("a after = {:?}", a);
    println!("b after = {:?}", b);
    println!("c after = {:?}", c);
}
```

这里创建了一个`Rc<RefCell<i32>>`实例，接着在`a`中用包含`value`的`Cons`成员创建了一个`List`。我们需要克隆`value`以便`a`和`value`都能拥有其内部值`5`的所有权，而不是将所用权从`value`移动到`a`或者让`a`借用`value`。

我们将`a`封装进`Rc<T>`，这样当创建`b`和`c`时，它们都可以引用`a`。

一旦创建了`a`，`b`和`c`，我们将`value`的值加 10。为此对`value`调用了`borrow_mut`方法，这里使用了第五章讨论过的自动解引用功能来解`Rc<T>`从而获取其内部的`RefCell<T>`值。`borrow_mut`方法返回`RefMut<T>`智能指针，可以对其使用解引用运算符来修改其内部值。

执行结果如下：

```null
$ cargo run
   Compiling cons-list v0.1.0 (file:///projects/cons-list)
    Finished dev [unoptimized + debuginfo] target(s) in 0.63s
     Running `target/debug/cons-list`
a after = Cons(RefCell { value: 15 }, Nil)
b after = Cons(RefCell { value: 3 }, Cons(RefCell { value: 15 }, Nil))
c after = Cons(RefCell { value: 4 }, Cons(RefCell { value: 15 }, Nil))
```

...
